Add autogenerated Spack constraints for Trilinos

kuberry opened the pull request:
This is a step towards addressing [#33760](https://github.com/spack/spack/issues/33760). It sets defaults to False for Trilinos package variants and adds subpackages of Trilinos as variants. 

The variants and conflicts are autogenerated using https://github.com/kuberry/spack/blob/tribits_spack_generation/spack_variants_from_tribits_projects.py 

This is not intended to be ready to merge at the moment. I'm creating this PR to demonstrate what the translation of Tribits logic to Spack will/can look like.

kuberry committed:
Add autogenerated Spack constraints for Trilinos

- added note about how to use new Trilinos spec

kuberry committed:
Add back in a few conflicts

bartlettroscoe reviewed (commented):
@kuberry, would it not be possible to write all of this variant and conflict into into a separate *.py or JSON file and then have the file read in by trilinos/package.py when it gets run?  That would keep the package.py file cleaner and make the updates of this information very clean.

bartlettroscoe commented:
And I might suggest that you just add the script:

* [kuberry/spack@tribits_spack_generation/spack_variants_from_tribits_projects.py](https://github.com/kuberry/spack/blob/tribits_spack_generation/spack_variants_from_tribits_projects.py?rgh-link-date=2022-11-08T17%3A34%3A48Z)

to the Spack repo under the trilinos/ directory right beside the trilinos/package.py file along with a README.md file for how to run it to update this info for new versions of Trilinos.

kuberry commented:
> @kuberry, would it not be possible to write all of this variant and conflict into into a separate *.py or JSON file and then have the file read in by trilinos/package.py when it gets run? That would keep the package.py file cleaner and make the updates of this information very clean.

Hi Ross, that seems very possible. I started this PR to collect that type of feedback.

kuberry committed:
Add Tribits->Spack translator

- all_optional_packages_diff.py added for determining the effect of
  adding +all_optional_packages
- added translator
- moved hints and tips on installation into package description

alalazo commented:
Closing, since there has been no activity in a while. Feel free to migrate the PR / reopen at https://github.com/spack/spack-packages if you want to continue work on this.
